velox/connectors/hive/storage_adapters/s3fs/RegisterS3FileSystem.cpp

```diff
@@ -26,30 +29,59 @@
 namespace facebook::velox::filesystems {
 
 #ifdef VELOX_ENABLE_S3
-folly::once_flag S3FSInstantiationFlag;
-
-// Only one instance of S3FileSystem is supported for now.
-// TODO: Support multiple S3FileSystem instances using a cache
-static std::shared_ptr<S3FileSystem> s3fs = nullptr;
+std::mutex mtx1, mtx2;
+/// Multiple S3 filesystems are supported.
+/// Key is the endpoint value specified in the config using hive.s3.endpoint.
+/// If the endpoint is empty, it will default to AWS S3.
+/// We cannot use folly::ConcurrentHashMap to store the filesystem instances.
+/// This is because we need to delete the filesystems during
+/// finalizeS3FileSystem and folly::ConcurrentHashMap does not guarantee the
+/// entries are cleared on invoking clear().
+static std::unordered_map<std::string, std::shared_ptr<FileSystem>> filesystems;
 
 std::function<std::shared_ptr<
     FileSystem>(std::shared_ptr<const Config>, std::string_view)>
 fileSystemGenerator() {
```

Comment:
maybe a dumb question, but why do we need function returning a function pointer? Could we just define this as a regular function, and wrap that in an std::function when we need a pointer to it?

Reply:
We can simply this. Fixed.